### DISCUSSION TITLE
Best Way to Safely Remove MP3Gain Tags from Station MP3 Files in AzuraCast

### DISCUSSION THREAD

**Discussion by BornKillaz** [AzuraCast/AzuraCast#7168] [Q&A]
Hello, everyone!

I'm currently running an AzuraCast instance to manage 1 radio station, and I've been encountering some playback issues with my MP3 files. Specifically, I've noticed errors such as:
`[mp3float @ 0x7f187b6b7800] Could not update timestamps for skipped samples`

Additionally, my jingles, which are set to play every 2 songs, are playing inconsistently, sometimes every 2, 3, or 4 songs. I suspect that MP3Gain tags might be causing these playback issues.

I want to safely remove MP3Gain tags from all the MP3 files in my station's media library to see if this resolves the playback issues. This will be a one-time operation, so I am considering a shell script to remove MP3Gain tags using mp3gain:

#!/bin/bash

MP3_DIR="/var/azuracast/stations/my_station_id/media"

for FILE in "$MP3_DIR"/*.mp3; do
    if [[ -f "$FILE" ]]; then
        echo "Processing: $FILE"
        mp3gain -s c "$FILE"
        echo "MP3Gain tags removed from $FILE"
    else
        echo "No MP3 files found in the directory."
    fi
done

I foresee a couple of challenges with this approach:

- Ensuring the script runs reliably within the Docker container.
- Avoiding any potential data loss or corruption of the MP3 files.

I'm reaching out to gather insights and opinions on the best practices for removing MP3Gain tags from MP3 files as a one-time operation. Are there any other tools or methods that might be more effective or safer for removing MP3Gain tags?

Any feedback, suggestions, or shared experiences would be greatly appreciated!

Thank you!

**Comment by gAlleb** [comment]
> I'm currently running an AzuraCast instance to manage 1 radio station, and I've been encountering some playback issues with my MP3 files. Specifically, I've noticed errors such as:
[mp3float @ 0x7f187b6b7800] Could not update timestamps for skipped samples

It's an internal ffmpeg log and I wouldn't worry about it at all.

Also I don't see how MP3Gain tags can interfere with your playback. Just leave them be.

> my jingles, which are set to play every 2 songs, are playing inconsistently, sometimes every 2, 3, or 4 songs. 

This could be caused by many circumstances. If you want complete accuracy, consider `Advanced playlists` for jingles.

**Comment by Moonbase59** [comment]
Well, you could use [loudgain’s strip tags feature](https://github.com/Moonbase59/loudgain?tab=readme-ov-file#strip-unwanted-id3v1apev2-tags--s---striptags) and `rgbpm` for mass tagging, but then you would recalculate ReplayGain again…

I also don’t believe that being the cause of your issue.

**Reply by BornKillaz** [reply]
Back into your suggestion, and after this [topic,](https://github.com/AzuraCast/AzuraCast/discussions/7171) I belive that would be the solution, since using Azuracast Reply Gain setting the gain on real-time, can cause issues. 

The concern is now on how to properly apply that into a long term solution.
My goal is to have that running like a cron on AzuraCast media files, looking for untagged files and  properly adding the Gain tag to those files... You know, sometimes people are really lazy and "forget" to properly do it before upload media to AzuraCast. 

Also,  a huge addition to AzuraCast would be an option to apply user defined gain on uploaded files, basically using the same method it uses to scan files to add them to folder based playlists. Don't know if this would ever be a reality.

**Reply by gAlleb** [reply]
1. If you have "mp3s" (I assume you do). You can disable replaygain option and turn on `autocue` option. (Don't forget to back up your Azuracast installation with all the music).

2. Then go to "Broadcasting" -> "Edit Liquidsoap Configuration" and add this lines into 2nd box and save:
```
settings.autocue.cue_file.noclip := true  # clipping prevention
settings.autocue.cue_file.write_tags := true
```

3. This way during playback `autocue` will analyze your songs and will write all the tags needed into the files, including `liq_amplify` (which is similar to ReplayGain but uses ffmpeg ebur128 filter). So eventually all of your tracks will be tagged and as a bonus you will have nice `cue_in/cue_out" values)) And next time those tracks will be played - the values won't be calculated again, thus much less CPU consumption. 


> to apply user defined gain

user defined gain? If I understand correctly and you want to use another target, different from -18 LUFS, you can add this to the settings above and it would use that LUFS target with "liq_amplify"
```
settings.autocue.cue_file.target := -18.0
```

**Reply by BornKillaz** [reply]
@gAlleb, this seems to be a solution for me, but if I understand it correctly, the first time a track is played, it will have no gain/LUFS adjustment. 

So, on first playback, whether it's low or high, it will sound as is and will only have the correct volume the next time it is played. 

Is that correct?

**Reply by gAlleb** [reply]
No, it's not ccorrect. The first time a track is played, it `will have` gain/LUFS adjustment.

It would have correct volume from the start. `Autocue` will analyze the track, normalize it, apply all settings to it and write tags as well on the first time it encounters "new" track. Playback will be already with applied LUFS target.
Second time it would apply already written tags - thus no computation, less CPU.

**Reply by BornKillaz** [reply]
That is handsome! 

Is there any way to avoid a new autocue analysis and tag writing if the file already has a specified LUFS value equal to the specified LUFS target? (In my case, it would be -17 LUFS).

If so, I could use Loudgain to apply my -17 LUFS value to large MP3 files uploaded by the DJs, saving some resources. I have an entire DJ panel that I built, where DJ's can upload their pre-recorded shows (1 or 2 hours long, depending on the show) to Azuracast, and I can process those files upon upload.

Just to be clear, does applying liq_amplify  LUFS re-encode the file, or does it just write MP3 tags for gain?

Thank you.

**Reply by gAlleb** [reply]
>  Is there any way to avoid a new autocue analysis and tag writing if the file already has a specified LUFS value equal to the specified LUFS target? (In my case, it would be -17 LUFS).

Files should be processed by `cue_file`, because it writes specific tags:

```
    liq_cue_duration: 95.90
    liq_cue_in      : 0.00
    liq_cue_out     : 95.90
    liq_cross_start_next: 95.90
    liq_longtail    : false
    liq_loudness    : -17.75 LUFS
    liq_loudness_range: 2.71 LU
    liq_amplify     : -0.25 dB
    liq_amplify_adjustment: 0.00 dB
    liq_reference_loudness: -18.00 LUFS
    liq_blankskip   : false
    liq_blank_skipped: false
    liq_true_peak   : -4.94 dBFS
```

And Loudgain writes specific `replaygain` tags:

```
    replaygain_album_gain: -2.61 dB
    replaygain_album_peak: 1.037845
    replaygain_album_range: 15.67 dB
    replaygain_reference_loudness: -18.00 LUFS
    replaygain_track_gain: -0.26 dB
    replaygain_track_peak: 0.566594
    replaygain_track_range: 2.62 dB
```

Which as you can see - are different. What you can do is download `cue_file` from here https://github.com/Moonbase59/autocue/tree/integrate-with-liquidsoap (make it executable, move it to `/usr/local/bin`, you need `Python3` installed on your system) and then write tags to your files before uploading them: `cue_file -w filename.mp3`. This is how you can process those files before upload. If you need -17 LUFS:

```
cue_file -w -t -17 --noclip filename.mp3
```
`-w` = write tags
`-t -17` = reference loudness
`--noclip` (or `-k`) = clipping prevention

I hope that soon enough the functionality of tagging new files upon upload will be available in Azuracast. 


Also beware, that your files nevertheless should have a `replaygain_track_gain` in them already with LUFS corresponding to your desired target `replaygain_reference_loudness`. If your files don't have this tags - autocue will compute the tracks everytime anyway. `So first you better Loudgain the tracks, then use `Autocue` with `write_tags` option`. That iis the best option because mandatory tags for skipping computation are:

```
# minimum set of tags after "read_tags" that must be there before skipping
# analysis
tags_mandatory = set([
    "duration",
    "liq_cue_in",
    "liq_cue_out",
    "liq_cross_start_next",
    "replaygain_track_gain",
])
```


> Just to be clear, does applying liq_amplify LUFS re-encode the file, or does it just write MP3 tags for gain?

It just writes new tags. To be more specific: it creates new temp file with tags added, then removes original file, and renames temp file. No re-encoding. You can download `cue_file` and play with it to see for yourself. It's the same `cue_file` which is used iin Azuracast https://github.com/AzuraCast/AzuraCast/blob/b0f11c8f2b71325cd6a2460c1fa59691cb2b5331/util/docker/stations/setup/autocue.sh#L9

**Reply by gAlleb** [reply]
Also to use -17 value - you need to add  `settings.autocue.cue_file.target := -17.0` into 2nd box of LS config. So `autocue` would write `liq_reference_loudness: -17.00 LUFS` into the tags.

**Reply by BornKillaz** [reply]
@gAlleb, Thank you very much for your valuable help. I will definitely try that approach. 👍

**Reply by gAlleb** [reply]
Just read through current Azuracast `cue_file` version again and looked at `mandatory tags`. So beware that in this version your files nevertheless should have a `replaygain_track_gain` in them already with LUFS corresponding to your desired target `replaygain_reference_loudness` to be able to skip computation. If your files don't have `replaygain_track_gain` tag - autocue will compute the tracks everytime anyway (bummer). So first you better Loudgain the tracks, then use Autocue with write_tags option. That is the best option because mandatory tags for skipping computation are:
```
# minimum set of tags after "read_tags" that must be there before skipping
# analysis
tags_mandatory = set([
    "duration",
    "liq_cue_in",
    "liq_cue_out",
    "liq_cross_start_next",
    "replaygain_track_gain",
])
```

So it seems you gotta `replaygain` your tracks nevertheless. However, you can wait a little. Maybe @Moonbase59 could come up with something.

**Reply by gAlleb** [reply]
@BornKillaz, @Moonbase59 did it! So now you don't have to pre-replaygain your tracks and can indeed follow the above instructions) Add `settings.autocue.cue_file.write_replaygain := true` to the 2nd box as well. And if using command: `cue_file -w -t -17 --noclip -r filename.mp3'

https://github.com/AzuraCast/AzuraCast/discussions/6252#discussioncomment-9669300

```txt
Version 2.0.0 is out! It offers the chance to write ReplayGain data to
audio files, in addition to the Liquidsoap/Autocue tags.

This is useful for station owners that can’t or don’t want to use audio tracks
that have been pre-replaygained. You can now add ReplayGain data to songs and
thus ensure the analysis won’t be done over and over again.

This is an additional option to the existing
write_tags feature and will only work when write_tags is enabled. Based on
our file analysis, it will save (and overwrite) ReplayGain Track data, such as
the gain value, the reference loudness, integrated loudness and true peak info.
```

Thanks, @Moonbase59!

**Reply by BornKillaz** [reply]
Thank you @gAlleb. I will gladly test this as soon as I can enable autocue option, becouse at this time, when I enable it, my station does not come online, it just keeps playing the fallback file over and over again. 

My only option to get my station back was to totally disable ReplayGain and of course, autocue. And on top of that, the smart crossfade is not working anymore.

So, hopefully, this new problems get to be solved so I can at least enable autocue or at least ReplayGain, as the station now sounds horrible.

**Reply by gAlleb** [reply]
The reason your station shows "offline" and plays fallback is because `autocue` has to process first songs on all of your playlists ( with `prefetch=2` option). While it does that - CPU spikes up - thus fallback is activated. You just have to wait a bit and then station would play. Considering that you've got 2 hour tracks in there - it indeed needs more time and if you'll open `htop` or `top` at this moment of startup you will see `ffmpeg` working hard! The reason behind that: autocue is implemented in a way that all of your playlists are  annotated now with `autocue protocol`. And even if they are not due to play right now - first few songs have to be processed nevertheless.

Just wait a bit. And station will go online. 

When Buster will update `autocue` with new config and `cue_file` - you can start using it as I have mentioned before. And then (if you'll switch on the `write_mode` - you'll start writing necessary tags into files. Once all of your files are with tags — start up of you station would take much much less time.

**Reply by BornKillaz** [reply]
But @gAlleb, even if I totally disable autocue and restart, station won't start. It keeps playing the fallback file and after a while it just stops and starts playing the white noise. The only solution at that point is to disable the old ReplayGain and restart the station again. But the Smart Crossfade does not work at all. 

So autocue and this issues must be related, as if you enable autocue, it disables smart crossfade. Then, by disabling autocue and restart, the Smart Cross never gets to work again. ReplayGain was working on previous version and also was Smart Crossfade, now I cannot get either to work. 

Hopefully I can rollback to version 20.0 until these issues get fixed.

**Reply by BusterNeece** [reply]
@BornKillaz is that because you have custom Liquidsoap code where you've customized AutoCue? AzuraCast should work perfectly fine with AutoCue turned off, and if it isn't, then we need to see logs explaining why that's happening.

**Reply by BornKillaz** [reply]
@BusterNeece, No, I totally removed the custom autocue settings from the Liquidsop configurations when I disabled autocue for the first time. I tried with and without those custom configs, but no luck.

**Here are the Liquidsoap logs:** https://www.dropbox.com/scl/fi/3kt0ratspwact4bwwlza3/liquidsoap-logs.txt?rlkey=fykzrkskqt31ol5mg765hknpp&st=812eviq5&dl=0

**And here is my current Liquidsoap config file:** https://www.dropbox.com/scl/fi/82hftqglov2wh64j3d5ii/liquidsoap-config.txt?rlkey=72w1kl6u5h9mv1ikq5mgqxt6t&st=an7zolzw&dl=0

Hope this helps. 

Thank you @Moonbase59, @BusterNeece  and @gAlleb in advance for your time and concern about this matter.

**Reply by Moonbase59** [reply]
The config shows you’re not using autocue, but the logs say you do.

From the log:
```
2024/06/04 20:07:08 [autocue.cue_file:3] Now autocueing: "/var/azuracast/stations/all_stars_portugal/media/SLOTS-PAINEL-ANIMADORES/[SLOT] Rock Sounds (5ª feira, 22h - 23h)/Rock-Sounds-com-Cat-Color-ZAoVpr.mp3"

...

2024/06/04 20:07:42 [autocue.cue_file:3] cue_file result for "/var/azuracast/stations/all_stars_portugal/media/SLOTS-PAINEL-ANIMADORES/[SLOT] Rock Sounds (5ª feira, 22h - 23h)/Rock-Sounds-com-Cat-Color-ZAoVpr.mp3": {"duration": 2850.7999999999997, "liq_cue_duration": 2850.7, "liq_cue_in": 0.0, "liq_cue_out": 2850.7, "liq_cross_start_next": 2850.6, "liq_longtail": false, "liq_loudness": "-9.21 LUFS", "liq_loudness_range": "8.30 LU", "liq_amplify": "-8.79 dB", "liq_amplify_adjustment": "0.00 dB", "liq_reference_loudness": "-18.00 LUFS", "liq_blankskip": false, "liq_blank_skipped": false, "liq_true_peak": "3.05 dBFS"}
```
So it analysed a 48 minute track in 34s, quite good.

Here something bad happened, something took _way_ too much time:
```
2024/06/04 20:07:42 [request:2] Time limit exceeded by 169.00 secs!
```
and this indicates you have ReplayGain computation enabled as well:
```
2024/06/04 20:07:42 [file.replaygain.7:3] Computing replay gain for "/var/azuracast/stations/all_stars_portugal/media/SLOTS-PAINEL-ANIMADORES/[SLOT] Rock Sounds (5ª feira, 22h - 23h)/Rock-Sounds-com-Cat-Color-ZAoVpr.mp3"
```
_That_ is real bad! If Autocue is used, the other ReplayGain should _always be off!_ Otherwise, as probably happened here, they "fight" for the same file and even might _both_ try to write to it. Chaos and broken music files almost guaranteed!

Besides, it doesn’t make much sense letting ReplayGain calculate the same values Autocue calculates. That’s just wasting double CPU load, and might have the risk of damaging the music files.

---

Another wild guess: Have you been using the "old", manual Autocue before (where you had to copy all that code into input box #2)?

If you had, and not completely removed the "old" Autocue _and_ the old `cue_file`, I’ve also seen stations _never_ starting correctly. Also check your `/var/azuracast/docker-compose.override.yml` file for a volume mapping like this:

```yml
services:
  web:
    volumes:
      - /var/azuracast/bin/cue_file:/usr/local/bin/cue_file
```
Should that still be in there, failure is guaranteed. Just remove the line with `cue_file` in it.

After this change, the Docker must be restarted, for instance like so:
```bash
docker-compose down
docker-compose up -d
```

**Reply by BornKillaz** [reply]
@Moonbase59, this log covers several situations and experiments taken on that time period. So sometimes AutoCue is on, and sometimes it's disable, and I tried basically everything in despair. The Liquidsoap config is the actual one. 

You see, I could never use both autocue and ReplayGain (I don't want that on the first place) and have the station online, far from that. I can't even have ReplayGain on. At this point, very honestly, I just want that it all comes working just like it was before AutoCue. 

If I turn on AutoCue, station keeps offline, If I turn off AutoCue and turn on ReplayGain, station keeps offline. 

For now, the only way I can have the station online is having ReplayGain off and  autocue off, and to crossfade I have to use Normal crossfade, as Smart Crossfade does not work anymore, basically. 

And no, I have never had cue_file before and there is no such reference in my docker-compose.override.yml

I really want to provide as much feedback as I can so this can get solved. If I can then use AutoCue, that would be perfect, but for now being able to use ReplyGain would be perfect for me! 

As soon as I can, I will to test diferent scenarios and provide a separate liquidsoap log and config file for each scenario. 

Thank you.

**Comment by BornKillaz** [comment]
I'm doing the most to exclude all possible causes, and @Moonbase59, I read a post where you advise not to use Mp3Gain, so I wanted to make sure that this issues are not caused by the Replay Gain tags generated by Mp3Gain.

And I'm doing the AzuraCast built in Replay Gain in real time with Liquidsoap, so I really don't need those tags on those files. 

Thank you both for your replies.

**Comment by Moonbase59** [comment]
### A few words of advice here, if I may:

- In general, it pays to have your music collection in good order _before_ starting a station. And _keep_ it that way. That means, replaygaining and/or running cue_file _once_ over the files. I know of at least one user who has written a little "wrapper" around cue_file for tagging multiple files/folders beforehand.
- Independent of what loudness target you use for broadcasting, I really _strongly_ recommend to target your music collection at **-18 LUFS**, which has been good and recommended practice for many years now. It corresponds to the earlier mp3gain setting of "89 dB", which some might still know about. Players also typically don’t care much about the stored `replaygain_loudness_reference` (cue_file is an exception, for a reason!) and just **assume** all files are tagged at -18 LUFS and blindy use only the `replaygain_track_gain` value! All hell and great chaos would break loose if everyone suddenly started tagging their files with their own personal reference loudness. Imagine going to a party where everyone brings their music, and one song would hurt your ears while the next was so silent that no one understands the lyrics! That’s exactly what we want to _avoid_, isn’t it?
- I put some efforts in autocue so it can use files tagged at -18 LUFS and recalculate for another broadcasting playout loudness on-the-fly _without_ having to re-analyse the file in most cases. For exactly the reason above.
- That said, auto-calculating replaygain data and the ability to store cueing data in tags for faster access "next time round" are strictly meant as _convenience features_, _not_ for normal everyday use over the years. These features _are_ useful if you run a station where others can upload, or in cases where a station downloads files to play from outside sources (news, syndication). In both cases  you probably don’t know about the exact state of the music.
- _Why_ do I generally recommend doing it once and for all _before?_
  - _CPU load, energy used:_ Having to re-analyse files on-the-fly _can_ lead to high CPU use (and thus, energy bills), especially for files that are large (show recordings!) or hard to calculate. When running on a small, say 1- or 2-CPU cloud VM, this can hit you.
  - _Incompleteness:_ If you auto-tag while broadcasting, you’ll never be sure that _all_ of your files are getting tagged—it can of course only tag those that actually _get played!_
  - _Disc/SSD wear and tear:_ Depending on file and tagging format, it may well be that the tagging software has to _rewrite the whole file_ (not in all cases), which puts some load on the disc, since files are constantly being rewritten. The system cannot _know_ if you have changed something in the meantime, so it’ll rewrite the tags _every time_ a song gets played.
  - _Write-protected, read-only files/folders/discs:_ You might have decided to keep your "sacred music" _safe_ (like I do), and made access read-only. If cue_file cannot write to a file, nothing bad happens, except it will have to do the analysis again each time, because the tags aren’t there.
  - _Higher-level software hits in:_ If you have playlists/files "on watch", which is typical in radio automation, that software might "think" files (you just rewrote) have been _changed_ and try to reprocess them, or reload the playlists. Example Liquidsoap: Auto-reloading playlists on watch. Example AzuraCast: Media handler auto-reprocessing changed files and updating database. Example Centovacast, LibreTime, and others: Same behaviour.
  - _Bandwidth:_ You might have to pay for bandwidth on cloud servers, and access your music remotely. Constant rewriting adds extra bandwidth being used.
- That said, for best performance, I’d advise to:
  - Find the "best" settings for your setup.
  - Include cue_file in your pre-processing chain (ideally after/in place of the ReplayGain step). That’s why I made it external, so you can use it for that purpose!
  - Upload pre-tagged files to your stations.
  - Immediate benefits!
    - No waiting on CPU-intensive analysis, all is ready for immediate playout.
    - Everything plays with perfect settings.
    - You save on energy bills, transfer volume cost, disc wear, time, and your nerves!
    - You can _still_ use things like AzuraCast’s Visual Cue Editor or Liquidsoap annotations to _override_ what’s been tagged in the files. Best of both worlds, really.
- Autocue/cue_file will still "fill the gaps", in case a few files aren’t pre-tagged, ensuring best playout possible.

A little planning before and using an optimized workflow has never been bad advice.

Now you can lean back and enjoy your station again, or maybe plan the next great show!

**Enjoy!**

**Reply by BornKillaz** [reply]
Thank you very much for your useful tips @Moonbase59. 

Can you point me out a windows software to add the ReplayGain tags to the mp3 files? The person that uploads the music into to the server, barely knows how to use a computer, so anything related to command line or SSH is totally not an option. He used to do that with Mp3Gain, but as you've pointed out, it's not recommended anymore.

Thank you.

**Reply by Moonbase59** [reply]
Well, I’d recommend my [`loudgain`](https://github.com/Moonbase59/loudgain) of course :smile:, but that would require running it in WSL2 (the Windows Subsystem for Linux).

Since `cue_file` is basically a Python3 script (I _knew_ why…) that in turn uses _Mutagen_, `ffmpeg` and `ffprobe`, for Windows (if it must be), I’d recommend:
- Install _Python3_.
- Install _ffmpeg_ (which will also bring _ffprobe_).
- For extra benefits, install _Mutagen_ (also Python). This can usually be done with `pip` or `pip3` like so: `pip3 install mutagen`. `cue_file` will auto-detect if it’s installed.
- Put Python3, ffmpeg, and `cue_file` in the systems’s path, so they can easily be called up in a terminal by just typing their name.

You could then use `cue_file` to pre-tag files on Windows, or even write a little Powershell script or the like to have it process whole folders. Use `cue_file --help` to get info about parameters, options, and supported file types and tags.

Haven’t been using Windows for 12+ years now, so let us know if it actually works on Windows. `cue_file` _should_ be cross-platform, but one never knows…

Maybe some other Windows user has a good recommendation, too? I know, for instance, that [_foobar2000_](https://www.foobar2000.org/) was a real good piece of software back in my Windows days. It has great tagging and also ReplayGain features.

_Hint:_ If settable, _don’t_ use the crappy _APEv2_ tags in MP3 files (personally, I’d remove those), but _only_ _ID3v2_ tags! I also recommend to _not_ store _ID3v1_ tags in files anymore—they’re outdated, limited, and only confuse playout software. Use _ID3v2.3_ or _ID3v2.4_, solely.

The _HydrogenAudio Knowledgebase_ also has a great [page on ReplayGain scanners](https://wiki.hydrogenaud.io/index.php?title=ReplayGain). It mentions _foobar2000_ as well as my _loudgain_.

**Reply by mishcazzulani** [reply]
Isn't it enough to add "settings.autocue.cue_file.write_tags := true" to the second box in liquidsoap settings in azuracast? 

This way autocue should run only once per file?

**Reply by Moonbase59** [reply]
@mishcazzulani It depends.

If you also want ReplayGain (track) data be stored, you _must_ also, in addition, use `write_replaygain`, like so:

```
settings.autocue.cue_file.write_tags := true
settings.autocue.cue_file.write_replaygain := true  # add RG (track) tags
```

This is mainly a security measure, since many users prefer to replaygain their files using other tools, and Autocue shouldn’t overwrite such information. This is why the default for `settings.autocue.cue_file.write_replaygain` is `false`.

Optionally, you can also activate clipping prevention (this info will also be stored, it affects the ReplayGain values):

```
settings.autocue.cue_file.noclip := true 
```

Autocue will still _run_ on every file, but immediately detect that it can skip the costly analysis, and be _much_ faster, using almost no extra CPU:

![grafik](https://github.com/AzuraCast/AzuraCast/assets/3706922/1421f634-51f2-4f8b-91c1-d88d72c1ed0d)

Also check my notes here: https://github.com/AzuraCast/AzuraCast/discussions/7168#discussioncomment-9673575

Another wild idea in this issue: https://github.com/AzuraCast/AzuraCast/issues/7175#issuecomment-2149370721

**Reply by mishcazzulani** [reply]
when adding settings.autocue.cue_file.write_replaygain := true i get this error:

At line 92, char 0-25: Error 5: this value has no method `write_replaygain` Its type is { blankskip : (() -> bool) .{comments : string, description : string, set : (bool) -> unit }, fade_in : (() -> float) .{comments : string, description : string, set : (float) -> unit }, fade_out : (() -> float) .{comments : string, description : string, set : (float) -> unit }, force_analysis : (() -> bool) .{comments : string, description : string, set : (bool) -> unit }, longtail : (() -> float) .{comments : string, description : string, set : (float) -> unit }, nice : (() -> bool) .{comments : string, description : string, set : (bool) -> unit }, noclip : (() -> bool) .{comments : string, description : string, set : (bool) -> unit }, overlay : (() -> float) .{comments : string, description : string, set : (float) -> unit }, overlay_longtail : (() -> float) .{comments : string, description : string, set : (float) -> unit }, path : (() -> string) .{comments : string, description : string, set : (string) -> unit }, silence : (() -> float) .{comments : string, description : string, set : (float) -> unit }, target : (() -> float) .{comments : string, description : string, set : (float) -> unit }, timeout : (() -> float) .{comments : string, description : string, set : (float) -> unit }, unify_loudness_correction : (() -> bool) .{comments : string, description : string, set : (bool) -> unit }, write_tags : (() -> bool) .{comments : string, description : string, set : (bool) -> unit } }.

**Reply by gAlleb** [reply]
@mishcazzulani Have your upgraded your Azuracast to latest RR?
https://github.com/AzuraCast/AzuraCast/commit/4cc26c544ce6a59f19d196a262962eb3bc98de42

**Reply by mishcazzulani** [reply]
no i'm on the stable branch v0.20.1 

write_tags and noclip works though!

**Reply by gAlleb** [reply]
Well, here is the answer.

**Reply by Moonbase59** [reply]
Yeah, should have mentioned that it is new in autocue v2.0.0, and works correctly since v2.0.1, see https://github.com/Moonbase59/autocue/blob/integrate-with-liquidsoap/CHANGELOG.md

This kind of message:
```
Error 5: this value has no method write_replaygain
```
is typically an indicator that some new function is not in your version yet.

Guess I’ll add a log message in the next version, which shows the autocue version you’re running in the log.

```
2024/06/05 16:16:03 [autocue.cue_file:2] You are using autocue.cue_file version 2.0.2.
2024/06/05 16:16:03 [autocue.cue_file:2] Assure that the external "cue_file" has the same version!
```
(Don’t ask—not yet released.)

**Comment by Warblefly** [comment]
Total agreement with this advice. Works very well for me.

**Comment by BornKillaz** [comment]
Just tried again. There is no way, no matter what, the station don't start if ReplayGain is active. Even when it starts, it plays one track from the current playlist, and goes back to the fallback mp3 file, forever.

So, AutoCue is of course turned off, I changed the cross fade into normal, since Smart Crossfade no longer works and ReplayGain off.

What's your advise @BusterNeece ? Should I update to Rolling Release version to test if it works, or that is going to make no difference?

If not, I'll really need to rollback to version 20.0, but, I'm not sure on how safe would be to rollback. Any advice?

Thank you.
